include/zephyr/drivers/sensor.h

```diff
@@ -394,12 +398,184 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
+	 * @return 0 on success
+	 * @return -EINVAL if the @p channel_type doesn't exist in the buffer
+	 * @return <0 on error
+	 */
+	int (*get_shift)(const uint8_t *buffer, enum sensor_channel channel_type, int8_t *shift);
+
+	/**
+	 * @brief Decode up to N samples from the buffer
+	 *
+	 * This function will never wrap frames. If 1 channel is available in the current frame and
+	 * @p max_count is 2, only 1 channel will be decoded and the frame iterator will be modified
+	 * so that the next call to decode will begin at the next frame.
+	 *
+	 * @param[in]     buffer The buffer provided on the :c:struct:`rtio` context
+	 * @param[in,out] fit The current frame iterator
+	 * @param[in,out] cit The current channel iterator
+	 * @param[out]    channels The channels that were decoded
+	 * @param[out]    values The scaled data that was decoded
+	 * @param[in]     max_count The maximum number of channels to decode.
+	 * @return
+	 */
+	int (*decode)(const uint8_t *buffer, sensor_frame_iterator_t *fit,
+		      sensor_channel_iterator_t *cit, enum sensor_channel *channels, q31_t *values,
+		      uint8_t max_count);
+};
+
+/**
+ * @typedef sensor_get_decoder_t
+ * @brief Get the decoder associate with the given device
+ *
+ * @see sensor_get_decoder for more details
+ */
+typedef int (*sensor_get_decoder_t)(const struct device *dev,
+				    const struct sensor_decoder_api **api);
+
+/*
+ * Internal data structure used to store information about the IODevice for async reading and
+ * streaming sensor data.
+ */
+struct sensor_read_config {
+	const struct device *sensor;
+	enum sensor_channel *const channels;
+	size_t num_channels;
+	const size_t max_channels;
+};
+
+/**
+ * @brief Define a reading instance of a sensor
+ *
+ * Use this macro to generate a :c:struct:`rtio_iodev` for reading specific channels. Example:
+ *
+ * @code(.c)
+ * SENSOR_DT_READ_IODEV(icm42688_accelgyro, DT_NODELABEL(icm42688),
+ *     SENSOR_CHAN_ACCEL_XYZ, SENSOR_CHAN_GYRO_XYZ);
+ *
+ * int main(void) {
+ *   sensor_read(&icm42688_accelgyro, &rtio);
+ * }
+ * @endcode
+ */
+#define SENSOR_DT_READ_IODEV(name, dt_node, ...)                                                   \
+	static enum sensor_channel __channel_array_##name[] = {__VA_ARGS__};                       \
+	static struct sensor_read_config __sensor_read_config_##name = {                           \
+		.sensor = DEVICE_DT_GET(dt_node),                                                  \
+		.channels = __channel_array_##name,                                                \
+		.num_channels = ARRAY_SIZE(__channel_array_##name),                                \
+		.max_channels = ARRAY_SIZE(__channel_array_##name),                                \
+	};                                                                                         \
+	RTIO_IODEV_DEFINE(name, &__sensor_iodev_api, &__sensor_read_config_##name)
+
+static inline void sensor_reconfigure_read_iodev(struct rtio_iodev *iodev,
```

Comment:
@teburd I want to get your take on this function, it is not currently used by the sensor_shell but it i think it needs to be when userspace is enabled.

Reply:
This seems fine, but would need to be made a syscall when you enable userspace.

Reply:
Correct

Reply:
Done

Reply:
@teburd I enabled userspace and ran the sensor shell sample with directly manipulating the array and it worked. My guess is that shell commands run in supervisor mode. Any thoughts of a use case we can commit for this?